### PR TITLE
chore: clean up lodash usage

### PR DESCRIPTION
**Pull request by fson**
- Removed every `import _ from 'lodash'` usage and replaced it with individual imports (which is the prevailing convention in this codebase). This way we don't end up bundling/parsing the whole library, when we only use a subset. This makes it also easier to later switch to individual `lodash.*` packages, in cases where doing that makes sense.
- Replaced most uses of `lodash/get` with plain TypeScript, because the return value of `get()` is `any`, which hides critical type errors. (Found a few potential errors this way.)
- Replaced `lodash/concat`, `lodash/every` etc, with plain TypeScript.

### CODE REVIEW

File: packages/expo-cli/src/commands/build/ios/IOSBuilder.ts [hunk 1]
```diff
@@ -336,7 +339,7 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
     }
   }
 
-  determineCredentialsToClear(): Dictionary<boolean> {
+  determineCredentialsToClear(): { [name: string]: boolean } {
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/build/ios/IOSBuilder.ts:342]:
```suggestion
  determineCredentialsToClear(): Record<string, boolean> {
```
nit

File: packages/expo-cli/src/commands/build/ios/IOSBuilder.ts [hunk 2]
```diff
@@ -301,7 +304,7 @@ See https://docs.expo.io/distribution/building-standalone-apps/#2-configure-appj
     ctx: Context,
     experienceName: string,
     bundleIdentifier: string,
-    credsToClear: Dictionary<boolean>
+    credsToClear: { [name: string]: boolean }
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/build/ios/IOSBuilder.ts:307]:
```suggestion
    credsToClear: Record<string, boolean>
```

File: packages/expo-cli/src/commands/client/index.ts [hunk 3]
```diff
@@ -61,19 +61,19 @@ export default function (program: Command) {
         }
         if (!exp.ios) exp.ios = {};
 
-        if (!_.has(exp, 'facebookAppId') || !_.has(exp, 'facebookScheme')) {
+        if (!has(exp, 'facebookAppId') || !has(exp, 'facebookScheme')) {
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/client/index.ts:64]:
```suggestion
        if (!exp.facebookAppId || !exp.facebookScheme) {
```
Maybe something like this?

File: packages/expo-cli/src/commands/client/index.ts [hunk 4]
```diff
@@ -61,19 +61,19 @@ export default function (program: Command) {
...
           const disabledReason = exp
             ? `facebookAppId or facebookScheme are missing from app configuration. `
             : 'No custom configuration file could be found. You will need to provide a json file with valid facebookAppId and facebookScheme fields.';
           disabledServices.facebookLogin = { name: 'Facebook Login', reason: disabledReason };
         }
-        if (!_.has(exp, 'ios.config.googleMapsApiKey')) {
+        if (!has(exp, 'ios.config.googleMapsApiKey')) {
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/client/index.ts:70]:
```suggestion
        if (!exp.ios?.config?.googleMapsApiKey) {
```
Maybe?

File: packages/expo-cli/src/commands/client/index.ts [hunk 5]
```diff
@@ -61,19 +61,19 @@ export default function (program: Command) {
...
           const disabledReason = exp
             ? `ios.config.googleMapsApiKey does not exist in the app configuration.`
             : 'No custom configuration file could be found. You will need to provide a json file with a valid ios.config.googleMapsApiKey field.';
           disabledServices.googleMaps = { name: 'Google Maps', reason: disabledReason };
         }
-        if (_.has(exp, 'ios.googleServicesFile')) {
+        if (has(exp, 'ios.googleServicesFile')) {
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/client/index.ts:76]:
```suggestion
        if (exp.ios.googleServicesFile) {
```

File: packages/expo-cli/src/commands/client/index.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { Android, Simulator, UserManager, Versions } from '@expo/xdl';
 import chalk from 'chalk';
 import CliTable from 'cli-table3';
 import fs from 'fs-extra';
-import _ from 'lodash';
+import has from 'lodash/has';
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/client/index.ts:5]: Prolly don't need this

File: packages/expo-cli/src/commands/upgrade.ts [hunk 7]
```diff
@@ -494,11 +496,11 @@ export async function upgradeAsync(
   let updates = await getUpdatedDependenciesAsync(projectRoot, workflow, targetSdkVersion);
 
   // Split updated packages by dependencies and devDependencies
-  let devDependencies = _.pickBy(updates, (_version, name) => _.has(pkg.devDependencies, name));
+  let devDependencies = pickBy(updates, (_version, name) => has(pkg.devDependencies, name));
   let devDependenciesAsStringArray = Object.keys(devDependencies).map(
     name => `${name}@${updates[name]}`
   );
-  let dependencies = _.pickBy(updates, (_version, name) => _.has(pkg.dependencies, name));
+  let dependencies = pickBy(updates, (_version, name) => has(pkg.dependencies, name));
```
Comment (EvanBacon) [inline packages/expo-cli/src/commands/upgrade.ts:503]:
```suggestion
  let dependencies = pickBy(updates, (_version, name) => pkg.dependencies[name]);
```
or 
```suggestion
  let dependencies = pickBy(updates, (_version, name) => name in pkg.dependencies);
```

File: packages/expo-cli/src/credentials/views/IosAppCredentials.ts [hunk 8]
```diff
@@ -1,12 +1,4 @@
-import chalk from 'chalk';
-import get from 'lodash/get';
-
 import { Context, IView } from '../context';
-import { IosAppCredentials, IosCredentials, IosPushCredentials } from '../credentials';
-import { askForUserProvided } from '../actions/promptForCredentials';
-import { displayIosUserCredentials } from '../actions/list';
-import prompt from '../../prompt';
-import log from '../../log';
```
Comment (EvanBacon) [inline packages/expo-cli/src/credentials/views/IosAppCredentials.ts:9]: dang

File: packages/expo-cli/src/credentials/views/Select.ts [hunk 9]
```diff
@@ -135,7 +134,7 @@ export class SelectAndroidExperience implements IView {
     this.askAboutProjectMode = false;
 
     if (this.androidCredentials.length === 0) {
-      this.androidCredentials = get(await ctx.api.getAsync('credentials/android'), 'credentials');
+      this.androidCredentials = (await ctx.api.getAsync('credentials/android')).credentials;
```
Comment (EvanBacon) [inline packages/expo-cli/src/credentials/views/Select.ts:137]:
```suggestion
      this.androidCredentials = (await ctx.api.getAsync('credentials/android'))?.credentials;
```

Maybe? Just guessing from the `get` command.

File: packages/xdl/src/Android.ts [hunk 10]
```diff
@@ -609,7 +613,7 @@ See https://docs.expo.io/guides/splash-screens/#splash-screen-api-limitations-on
     return;
   }
 
-  if (_.some(androidSplashImages, ({ sizeMatches }) => !sizeMatches)) {
+  if (some(androidSplashImages, ({ sizeMatches }) => !sizeMatches)) {
```
Comment (EvanBacon) [inline packages/xdl/src/Android.ts:616]:
```suggestion
  if (androidSplashImages.some({ sizeMatches }) => !sizeMatches)) {
```
maybe use built-in.

File: packages/xdl/src/detach/AndroidIntentFilters.ts [hunk 11]
```diff
@@ -26,7 +26,7 @@ export default function renderIntentFilters(intentFilters: any) {
 }
 
 function renderIntentFilterDatumEntries(datum: any) {
-  return _.toPairs(datum)
+  return toPairs(datum)
```
Comment (EvanBacon) [inline packages/xdl/src/detach/AndroidIntentFilters.ts:29]:
```suggestion
  return Object.entries(datum)
```
Hard to tell from the `any` types, but prolly can use `Object.entries`.

File: packages/xdl/src/detach/AndroidShellApp.js [hunk 12]
```diff
@@ -132,9 +132,9 @@ function getRemoteOrLocalUrl(manifest, key, isDetached) {
   // in detached apps, `manifest` is actually just app.json, so there are no remote url fields
   // we should return a local url starting with file:// instead
   if (isDetached) {
-    return _.get(manifest, key);
+    return get(manifest, key);
```
Comment (EvanBacon) [inline packages/xdl/src/detach/AndroidShellApp.js:135]:
```suggestion
    return manifest?.[key];
```

File: packages/xdl/src/detach/AndroidShellApp.js [hunk 13]
```diff
@@ -132,9 +132,9 @@ function getRemoteOrLocalUrl(manifest, key, isDetached) {
...
   }
-  return _.get(manifest, `${key}Url`);
+  return get(manifest, `${key}Url`);
```
Comment (EvanBacon) [inline packages/xdl/src/detach/AndroidShellApp.js:137]:
```suggestion
  return manifest?.[`${key}Url`];
```

File: packages/xdl/src/detach/AndroidShellApp.js [hunk 14]
```diff
@@ -719,7 +715,7 @@ export async function runShellAppModificationsAsync(context, sdkVersion, buildMo
   }
 
   // Add app-specific intent filters
-  const intentFilters = _.get(manifest, 'android.intentFilters');
+  const intentFilters = get(manifest, 'android.intentFilters');
```
Comment (EvanBacon) [inline packages/xdl/src/detach/AndroidShellApp.js:718]:
```suggestion
  const intentFilters = manifest?.android?.intentFilters;
```

File: packages/xdl/src/detach/AndroidShellApp.js [hunk 15]
```diff
@@ -1355,7 +1351,7 @@ async function buildShellAppAsync(context, sdkVersion, buildType, buildMode) {
     });
     await fs.copy(
       outputPath,
-      _.get(context, 'build.android.outputFile') || `/tmp/shell-debug.${ext}`
+      get(context, 'build.android.outputFile') || `/tmp/shell-debug.${ext}`
```
Comment (EvanBacon) [inline packages/xdl/src/detach/AndroidShellApp.js:1354]:
```suggestion
      context?.build?.android?.outputFile || `/tmp/shell-debug.${ext}`
```

File: packages/xdl/src/detach/ExponentTools.ts [hunk 16]
```diff
@@ -125,8 +126,8 @@ async function spawnAsync(
 
 function createSpawner(buildPhase: string, logger?: Logger) {
   return (command: string, ...args: any[]) => {
-    const lastArg = _.last(args);
-    const optionsFromArg = _.isObject(lastArg) ? args.pop() : {};
+    const lastArg = last(args);
```
Comment (EvanBacon) [inline packages/xdl/src/detach/ExponentTools.ts:129]:
```suggestion
    const { length, [length - 1]: lastArg } = args;
```
or 

```js
const lastArg = args[args.length - 1];
```

File: packages/xdl/src/detach/IosIPABuilder.js [hunk 17]
```diff
@@ -124,8 +125,8 @@ export default function createIPABuilder(buildParams) {
       }
     );
     const plistRaw = output.join('');
-    const plistData = _.attempt(plist.parse, plistRaw);
-    if (_.isError(plistData)) {
+    const plistData = attempt(plist.parse, plistRaw);
```
Comment (EvanBacon) [inline packages/xdl/src/detach/IosIPABuilder.js:128]:
```suggestion
    const plistData = typeof plist.parse === 'function' ? plist.parse(plistRaw) : undefined;
```
Something like this?

File: packages/expo-cli/src/commands/build/ios/IOSBuilder.ts (hunk 1 above)
Comment (fson) [reply packages/expo-cli/src/commands/build/ios/IOSBuilder.ts:342]: sure

File: packages/expo-cli/src/credentials/views/Select.ts (hunk 9 above)
Comment (fson) [reply packages/expo-cli/src/credentials/views/Select.ts:137]: Good call. The return value was `any` so it was masking the error 👿

Comment (fson) [reply packages/expo-cli/src/credentials/views/Select.ts:137]: Although I'm still not sure if it's necessary... might just push the error down the line ...

File: packages/xdl/src/Android.ts (hunk 10 above)
Comment (fson) [reply packages/xdl/src/Android.ts:616]: Thanks, missed this earlier.

File: packages/xdl/src/detach/AndroidShellApp.js (hunk 12 above)
Comment (fson) [reply packages/xdl/src/detach/AndroidShellApp.js:135]: Yeah, I though of this too, but then realized this file is JS 😮 🙀

File: packages/xdl/src/detach/AndroidShellApp.js (hunk 13 above)
Comment (fson) [reply packages/xdl/src/detach/AndroidShellApp.js:137]: This is JS

File: packages/xdl/src/detach/AndroidShellApp.js (hunk 12 above)
Comment (EvanBacon) [reply packages/xdl/src/detach/AndroidShellApp.js:135]: ☠️

File: packages/xdl/src/detach/IosIPABuilder.js (hunk 17 above)
Comment (fson) [reply packages/xdl/src/detach/IosIPABuilder.js:128]: Went with 
```ts
    try {
      return plist.parse(plistRaw);
    } catch (error) {
      throw new Error(`Error when parsing plist: ${error.message}`);
    }
```
`attempt` followed by `isError` must be the weirdest way to catch errors I've seen this week 😄

**Pull request merged**